File: trees/has_path_sum.rb
require_relative 'binary_trees'

# Problem: Given a binary tree and a number, return true if the tree has a root-to-leaf path
# such that adding up all the values along the path equals the given number.
# Return false if no such path can be found.

# Example, for the binary tree below,
#      1
#    /    \
#   2      3
#  / \    /
# 4   5  6
#       /
#      8
# the root to leaf path sums are: 7 => 1 2 4; 8 => 1 2 5; 18 => 1 3 6 8
# The function will return true for numbers 7, 8, or 18, and false otherwise

def has_path_sum?(root, sum)
  return sum == 0 if !root
  remaining_sum = sum - root.value
  if !root.left_child && !root.right_child
    return remaining_sum == 0
  elsif root.left_child && root.right_child
    return has_path_sum?(root.left_child, remaining_sum) || has_path_sum?(root.right_child, remaining_sum)
  elsif root.left_child
    return has_path_sum?(root.left_child, remaining_sum)
  else
    return has_path_sum?(root.right_child, remaining_sum)
  end
end

h = Node.new(8)
f = Node.new(6, h)
e = Node.new(5)
d = Node.new(4)
c = Node.new(3, f)
b = Node.new(2, d, e)
a = Node.new(1, b, c)

root = a

p has_path_sum?(root, 7) == true
p has_path_sum?(root, 18) == true
p has_path_sum?(root, 1) == false
p has_path_sum?(h, 8) == true
p has_path_sum?(nil, 0) == true
p has_path_sum?(c, 17) == true

File: leetcode/java/prob112HasPathSum/HasPathSum.java
/*
Path Sum

Given a binary tree and a sum, determine if the tree has a root-to-leaf path
such that adding up all the values along the path equals the given sum.

For example:
Given the below binary tree and sum = 22,
              5
             / \
            4   8
           /   / \
          11  13  4
         /  \      \
        7    2      1
return true, as there exist a root-to-leaf path 5->4->11->2 which sum is 22.

*/

public class HasPathSum {

  private static class TreeNode {
    int val;
    TreeNode left, right;

    TreeNode(int x) {
      val = x;
    }
  }

  public static boolean hasPathSum(TreeNode root, int sum) {
    if (root == null) {
      return false;
    } else if (root.left == null && root.right == null) {
      return (root.val == sum);
    } else {
      int newSum = sum - root.val;
      return ( hasPathSum(root.left, newSum) || hasPathSum(root.right, newSum) );
    }
  }

  public static void main(String[] args) {
    /*
          5
         / \
        4   8
       /   / \
      11  13  4
     /  \      \
    7    2      1

    */
    TreeNode node1 = new TreeNode(5);
    TreeNode node2 = new TreeNode(4);
    TreeNode node3 = new TreeNode(8);
    TreeNode node4 = new TreeNode(11);
    TreeNode node5 = new TreeNode(13);
    TreeNode node6 = new TreeNode(4);
    TreeNode node7 = new TreeNode(7);
    TreeNode node8 = new TreeNode(2);
    TreeNode node9 = new TreeNode(1);

    node1.left = node2;
    node1.right = node3;

    node2.left = node4;

    node3.left = node5;
    node3.right = node6;

    node4.left = node7;
    node4.right = node8;

    node6.right = node9;

    System.out.println("hasPathSum(null, 0) = " + hasPathSum(null, 0));
    System.out.println("hasPathSum(null, 1) = " + hasPathSum(null, 1));
    System.out.println("hasPathSum(node1, 27) = " + hasPathSum(node1, 27));
    System.out.println("hasPathSum(node1, 22) = " + hasPathSum(node1, 22));
    System.out.println("hasPathSum(node1, 26) = " + hasPathSum(node1, 26));
    System.out.println("hasPathSum(node1, 18) = " + hasPathSum(node1, 18));
    System.out.println("hasPathSum(node1, 0) = " + hasPathSum(node1, 0));
    System.out.println("hasPathSum(node9, 1) = " + hasPathSum(node9, 1));
    System.out.println("hasPathSum(node9, 2) = " + hasPathSum(node9, 2));
  }
}